instrumentation/mongo/src/main/java/brave/mongo/TraceMongoCommandListener.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2013-2020 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package brave.mongo;
+
+import brave.Span;
+import brave.Tracer;
+import com.mongodb.event.CommandFailedEvent;
+import com.mongodb.event.CommandListener;
+import com.mongodb.event.CommandStartedEvent;
+import com.mongodb.event.CommandSucceededEvent;
+import org.bson.BsonDocument;
+import org.bson.BsonValue;
+import org.bson.json.JsonWriterSettings;
+
+import java.net.InetSocketAddress;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * A Mongo command listener that will report via Brave how long each command takes.
+ *
+ * <p>To use it, call <code>.addCommandListener(TraceMongoCommandListener.builder().tracer(tracer).build())</code>
+ * on the {@link com.mongodb.MongoClientOptions} or {@link com.mongodb.MongoClientSettings} object that is used to
+ * create the {@code MongoClient} to be instrumented.
+ */
+public class TraceMongoCommandListener implements CommandListener {
+  private final Tracer tracer;
+  private final int maxAbbreviatedCommandLength;
+  private final JsonWriterSettings jsonWriterSettings;
+  private final Map<Integer, Span> activeSpans = new ConcurrentHashMap<>();
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  protected TraceMongoCommandListener(final Builder builder) {
+    tracer = Objects.requireNonNull(builder.tracer, "tracer cannot be null");
```

Comment:
our policy is to not escalate the java version needed by the underlying driver. Please revert use of java 8 apis. anyway null checking should happen before here.